pool/app/session/entity.ml

```diff
@@ -424,41 +424,65 @@ module Calendar = struct
   type location =
     { id : Pool_location.Id.t
     ; name : Pool_location.Name.t
-    ; url : string
     }
   [@@deriving eq, show, yojson]
 
   type links =
-    { show_experiment : bool
-    ; show_session : bool
-    ; show_location_session : bool
-    ; experiment : string
-    ; session : string
-    ; location_session : string
+    { experiment : string option
+    ; session : string option
     }
   [@@deriving eq, show, yojson]
 
-  let create_links
-        ?(show_experiment = false)
-        ?(show_session = false)
-        ?(show_location_session = false)
-        experiment_id
-        session_id
-        ({ url; _ } : location)
-    =
-    let session_id = Id.value session_id in
+  let make_links actor guardian experiment_id session_id (location : location) : links =
+    let open Guard in
+    let externalize_opt = CCOption.map Sihl.Web.externalize_path in
+    let id_string = Uuid.Target.to_string in
+    let session_id = Uuid.target_of Id.value session_id in
+    let experiment_id = Uuid.target_of Experiment.Id.value experiment_id in
+    let location_id = Uuid.target_of Pool_location.Id.value location.id in
+    let check_guardian model target =
+      let open ValidationSet in
+      Persistence.PermissionOnTarget.validate_set
+        guardian
+        Pool_message.Error.authorization
+        (one_of_tuple (Permission.Read, model, Some target))
+        actor
+      |> CCResult.is_ok
+    in
+    let open Format in
+    let access_experiment = check_guardian `Experiment experiment_id in
     let experiment =
-      Format.asprintf "admin/experiments/%s" (Experiment.Id.value experiment_id)
-      |> Sihl.Web.externalize_path
+      externalize_opt
+      @@
+      match access_experiment with
+      | true -> Some (asprintf "/admin/experiments/%s" (id_string experiment_id))
+      | false -> None
     in
-    let session base_url = Format.asprintf "%s/sessions/%s" base_url session_id in
-    { show_experiment
-    ; show_session
-    ; show_location_session
-    ; experiment
-    ; session = session experiment
-    ; location_session = session url
-    }
+    (* TODO: Having access to the experiment gives access to the session, as well?
```

Comment:
In all current combinations, If someone has access to the Experiment, the one has also at least Read access to the sessions.